client/lib/user-settings/index.js

```diff
@@ -260,26 +260,46 @@ UserSettings.prototype.getSetting = function( settingName ) {
 UserSettings.prototype.updateSetting = function( settingName, value ) {
 	if ( has( this.settings, settingName ) ) {
 		set( this.unsavedSettings, settingName, value );
-
+		let canDeleteUnsavedSetting = false;
 		/*
 		 * If the two match, we don't consider the setting "changed".
 		 * user_login is a special case since the logic for validating and saving a username
 		 * is more complicated.
 		 */
 		if (
 			get( this.settings, settingName ) === get( this.unsavedSettings, settingName ) &&
-			'user_login' !== settingName
+			'user_login' !== settingName &&
+			'language' !== settingName
 		) {
+			canDeleteUnsavedSetting = true;
+		}
+
+		/*
+		 * language is a special case since we have to check for changes to locale_variant
+		 * this might be easier if we tracked the lang_id instead as we do in client/my-sites/site-settings/form-general.jsx
+ 		*/
+		if ( 'language' === settingName ) {
```

Comment:
This is a little convoluted but I'll try to explain the logic behind checks in this block:

1. If the language setting value is the same as the one already saved and it doesn't have a locale variant, it means this is the root language and we're okay to ignore the change. If our saved settings did contain a locale variant then we can assume that the user is trying to switch back to the root language
2. If the language setting value is the locale variant itself, we are also safe to ignore the change since we know that this is the currently saved locale.

There's probably a better way to do this check. One way would be to save and check against the `lang_id`, which is what we do in the general settings language picker. The locale variants have a different id to their parent languages so there'd be no need to keep track 2 x settings values.